include/envoy/http/header_map.h

```diff
@@ -191,25 +197,16 @@ class HeaderString {
   bool operator!=(absl::string_view rhs) const { return getStringView() != rhs; }
 
 private:
-  union Buffer {
-    // This should reference inline_buffer_ for Type::Inline.
-    char* dynamic_;
-    const char* ref_;
-  } buffer_;
-
-  // Capacity in both Type::Inline and Type::Dynamic cases must be at least MinDynamicCapacity in
-  // header_map_impl.cc.
-  union {
-    char inline_buffer_[128];
-    // Since this is a union, this is only valid for type_ == Type::Dynamic.
-    uint32_t dynamic_capacity_;
-  };
-
-  void freeDynamic();
+  enum class Type { Reference, Inline };
+
+  absl::variant<absl::string_view, absl::InlinedVector<char, 128>> buffer_;
+
   bool valid() const;
 
-  uint32_t string_length_;
-  Type type_;
+  /**
+   * @return the type of backing storage for the string.
+   */
+  Type type() const { return Type(buffer_.index()); }
```

Comment:
Can you add a comment above near Type/buffer_ that this function relies on the order of the enum matching the variant order? Potentially fragile? Potentially add some assert around this? Up to you on the assert.

Reply:
I like the idea of the assert, should make it more robust. Will add it.